### PR TITLE
Add ServiceBase

### PR DESCRIPTION
**Pull request by weshaggard**
This change add ServiceBase and some of the missing ServiceController APIs that are in .NET Framework.

resolves https://github.com/dotnet/corefx/issues/6024

The ported ServiceBase matches desktop except the dependency on EventLog is removed as we don't have EventLog support in .NET Core (at least not yet). 

The port doesn't include ServiceInstaller or ServiceProcessInstaller as we don't want to pull forward the entire Installer abstraction into .NET Core. So while we have ServiceBase a .NET Core hosted service will need to be installed/removed using other methods (sc.exe, installers, or via code like our test service). 

As part of the testing work I removed our dependency on the old native service and started using our new managed service which uses ServiceBase. 

PTAL @stephentoub @ericstj 

@TravisTheTechie @dasMulli it would be great if you guys could take a look as well.

### CODE REVIEW

**Comment by weshaggard** [comment]
test NETFX x86 Release Build Outerloop
test Windows x64 Debug Build Outerloop

**Comment by weshaggard** [comment]
test Outerloop NETFX x86 Release Build
test Outerloop Windows x64 Debug Build

File: src/Common/src/Interop/Windows/advapi32/Interop.SERVICE_DELAYED_AUTOSTART_INFO.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+internal partial class Interop
+{
+    internal partial class Advapi32
+    {
+        [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
+        public struct SERVICE_DELAYED_AUTOSTART_INFO
+        {
+            public bool fDelayedAutostart;
+        };
```
Comment (justinvp) [inline src/Common/src/Interop/Windows/advapi32/Interop.SERVICE_DELAYED_AUTOSTART_INFO.cs]: Nit: Unnecessary semicolon

File: src/Common/src/Interop/Windows/advapi32/Interop.SERVICE_DESCRIPTION.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+internal partial class Interop
+{
+    internal partial class Advapi32
+    {
+        [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
+        public struct SERVICE_DESCRIPTION
+        {
+            public IntPtr description;
+        };
```
Comment (justinvp) [inline src/Common/src/Interop/Windows/advapi32/Interop.SERVICE_DESCRIPTION.cs]: Nit: Unnecessary semicolon

File: src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs [hunk 3]
```diff
@@ -0,0 +1,910 @@
...
+        {
+            if (serviceName == null)
+                return false;
+
+            // not too long and check for empty name as well.
+            if (serviceName.Length > ServiceBase.MaxNameLength || serviceName.Length == 0)
+                return false;
+
+            // no slashes or backslash allowed
+            foreach (char c in serviceName.ToCharArray())
```
Comment (justinvp) [inline src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs]: `ToCharArray()` unnecessarily allocates a char array

File: src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs [hunk 4]
```diff
@@ -0,0 +1,910 @@
...
+                    }
+                }
+
+                // If we are being hosted, then Run will not have been called, since the EXE's Main entrypoint is not called.
+                if (!_initialized)
+                {
+                    Initialize(true);
+                }
+
+                if (Environment.OSVersion.Version.Major >= 5)
```
Comment (justinvp) [inline src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs]: Is this check still needed?

File: src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs [hunk 5]
```diff
@@ -0,0 +1,910 @@
...
+                }
+
+                _status.controlsAccepted = _acceptedCommands;
+                _commandPropsFrozen = true;
+                if ((_status.controlsAccepted & AcceptOptions.ACCEPT_STOP) != 0)
+                {
+                    _status.controlsAccepted = _status.controlsAccepted | AcceptOptions.ACCEPT_SHUTDOWN;
+                }
+
+                if (Environment.OSVersion.Version.Major < 5)
```
Comment (justinvp) [inline src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs]: Is this still needed?

File: src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceController.cs [hunk 6]
```diff
@@ -132,6 +137,22 @@ public string DisplayName
                     GenerateNames();
                 return _displayName;
             }
+            set
+            {
+                if (value == null)
+                    throw new ArgumentNullException("value");
```
Comment (justinvp) [inline src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceController.cs]: `nameof(value)`

File: src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceController.cs [hunk 7]
```diff
@@ -132,6 +137,22 @@ public string DisplayName
...
+
+                if (string.Compare(value, _displayName, StringComparison.OrdinalIgnoreCase) == 0)
```
Comment (justinvp) [inline src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceController.cs]: `string.Equals`

File: src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceController.cs [hunk 8]
```diff
@@ -203,6 +224,22 @@ public string MachineName
             {
                 return _machineName;
             }
+            set
+            {
+                if (!CheckMachineName(value))
+                    throw new ArgumentException(SR.Format(SR.BadMachineName, value));
+
+                if (string.Compare(_machineName, value, StringComparison.OrdinalIgnoreCase) == 0)
```
Comment (justinvp) [inline src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceController.cs]: `string.Equals`

File: src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceController.cs [hunk 9]
```diff
@@ -214,6 +251,26 @@ public string ServiceName
                     GenerateNames();
                 return _name;
             }
+            set
+            {
+                if (value == null)
+                    throw new ArgumentNullException("value");
```
Comment (justinvp) [inline src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceController.cs]: `nameof(value)`

File: src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceController.cs [hunk 10]
```diff
@@ -214,6 +251,26 @@ public string ServiceName
...
+
+                if (string.Compare(value, _name, StringComparison.OrdinalIgnoreCase) == 0)
```
Comment (justinvp) [inline src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceController.cs]: `string.Equals`

File: src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceProcessDescriptionAttribute.cs [hunk 11]
```diff
@@ -0,0 +1,45 @@
+//------------------------------------------------------------------------------
+// <copyright file="ServiceProcessDescriptionAttribute.cs" company="Microsoft">
+//     Copyright (c) Microsoft Corporation.  All rights reserved.
+// </copyright>                                                                
+//------------------------------------------------------------------------------
+
```
Comment (justinvp) [inline src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceProcessDescriptionAttribute.cs]: Incorrect license header

File: src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceProcessDescriptionAttribute.cs [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+namespace System.ServiceProcess {
+
+
+    using System;
+    using System.ComponentModel;
```
Comment (justinvp) [inline src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceProcessDescriptionAttribute.cs]: Run the codeformatter? Usings shouldn't be nested.

File: src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/SessionChangeDescription.cs [hunk 13]
```diff
@@ -0,0 +1,67 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.ServiceProcess
+{
+    public struct SessionChangeDescription
+    {
+        private SessionChangeReason _reason;
+        private int _id;
```
Comment (justinvp) [inline src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/SessionChangeDescription.cs]: Nit: `readonly`

File: src/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestServiceInstaller.cs [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+                using (ServiceController svc = new ServiceController(ServiceName))
+                {
+                    if (svc.Status != ServiceControllerStatus.Stopped)
+                    {
+                        svc.Stop();
+                        svc.WaitForStatus(ServiceControllerStatus.Stopped, TimeSpan.FromSeconds(30));
+                    }
+                }
+            }
+            //catch { }
```
Comment (dasMulli) [inline src/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestServiceInstaller.cs]: I guess this should be removed or commented back in (along with `//try`)

File: src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceProcessDescriptionAttribute.cs (hunk 11 above)
Comment (weshaggard) [inline src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceProcessDescriptionAttribute.cs]: Thanks for pointing out. I decided to not include this file at all but I forget to remove it from my original commit.

File: src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs (hunk 4 above)
Comment (weshaggard) [inline src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs]: Removed.

File: src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs (hunk 5 above)
Comment (weshaggard) [inline src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs]: removed.

**Comment by weshaggard** [comment]
@dotnet/dnceng what is the best way to determine which tests failed? I see https://mc.dot.net/#/user/weshaggard/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/57c16a39b435caccc19aa518ab8b527d98da83e9 but don't see any failures just some "Unknown" blocks. This is from https://ci3.dot.net/job/dotnet_corefx/job/master/job/windows-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_true_prtest/24/.

**Comment by MattGal** [comment]
@weshaggard  the ?s are actually tests that didn't run at all for that configuration (no matching work item) but til recently we had a bug that showed them as ?.

I'll take a look since it's definitely showing runs that I'd label as complete and passed as failures, then hand off to the FR team as needed.

**Comment by MattGal** [comment]
@weshaggard this is freaky, there's some kind of MC bug.  The failing test in all cases of those 4 runs is System.Runtime.Performance.Tests, which is timing out.  Example log [here](https://dotnetjobresults.blob.core.windows.net/build-aaed332ed0b74fe6900c55f4d04133b0/Windows.7.Amd64.Open/System.Runtime.Performance.Tests/output.log?sv=2015-04-05&sr=c&sig=IscWP8G1R1oSwFnBCH2VyxWmt4wPKv8OjlKx3pmYuPE%3D&st=2017-08-03T17%3A43%3A56.3724012Z&se=2017-09-02T17%3A43%3A56.3724012Z&sp=r).

I will raid a FR issue and tag you / Chad.  There's something definitely wrong about how this is getting rendered (I'm guessing there have been several builds of this, one of which succeeded, and stomped finding that)

**Comment by weshaggard** [comment]
>(I'm guessing there have been several builds of this, one of which succeeded, and stomped finding that)

Yes that is why I have an open request to split the results across the different legs.

**Comment by weshaggard** [comment]
test Linux x64 Release Build

File: src/System.ServiceProcess.ServiceController/ref/Configurations.props [hunk 15]
```diff
@@ -2,8 +2,9 @@
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
     <BuildConfigurations>
-      netstandard;
+      netcoreapp;
```
Comment (ericstj) [inline src/System.ServiceProcess.ServiceController/ref/Configurations.props]: Why was this needed?

File: src/System.ServiceProcess.ServiceController/src/Configurations.props [hunk 16]
```diff
@@ -2,7 +2,7 @@
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
     <BuildConfigurations>
-      netstandard-Windows_NT;
+      netcoreapp-Windows_NT;
```
Comment (ericstj) [inline src/System.ServiceProcess.ServiceController/src/Configurations.props]: This means UWP no longer gets a valid impl.  I guess that's OK.  I doubt any of the API would work from appcontainer.

Comment (weshaggard) [inline src/System.ServiceProcess.ServiceController/src/Configurations.props]: correct UWP isn't supporting these. On top of the win32 APIs it requires admin access for most of these actions.

File: src/System.ServiceProcess.ServiceController/ref/Configurations.props (hunk 15 above)
Comment (weshaggard) [inline src/System.ServiceProcess.ServiceController/ref/Configurations.props]: No I will remove it and just leave netstandard and netfx given we have to have a netstandard ref now for compat with the older package.
